### DISCUSSION TITLE
Miri complains about `tokio::select`

### DISCUSSION THREAD

**Discussion by soerenmeier** [tokio-rs/tokio#3882] [Q&A] [answered]
While running some tokio code through miri it complained about: `no item granting read access to tag <260290> at alloc104175+0x2d9 found in borrow stack.`
Code: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1941d901dbe003cc292c9dac8963008b).
After expanding the code miri pointed to `disabled`.
```rust
if disabled & mask == mask {
    continue;
}
```
Expanded Code: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=751c416461d1377a078053150c28f463).
To run miri I'm using the following command: `MIRIFLAGS="-Zmiri-disable-isolation" cargo +nightly miri run`.

Is this a bug? And if yes in tokio or miri?

**Comment by Darksonn** [comment, marked as answer]
This most likely has to do with the situation described [here](https://gist.github.com/Darksonn/1567538f56af1a8038ecc3c664a42462). Expanding on that, it is likely related to timers, not `select!`.

**Reply by soerenmeier** [reply]
Thanks for the quick response and the link that really describes the self referential 'problem' well, even if i haven't understood it fully, need to read it again.
By timers, you mean inside the runtime?
It seems the error is only returned if channel or Sleep returns Poll::Ready.
What i find odd is that miri points to `disabled`.
Will play around with it some more.

**Reply by Darksonn** [reply]
By timer I mean that `tokio::time::Sleep` is part of an intrusive list.

**Reply by soerenmeier** [reply]
After replacing sleep with another channel the error still appeared. So channels are probably also part of an intrusive list.

**Reply by Darksonn** [reply]
It might be due to the unsoundness in async/await syntax itself, which you can read about in the first link in the article.

**Reply by soerenmeier** [reply]
Interesting after reading the thread about that i though they made it safe by removing noalias in some places. But miri maybe doens't know about that.
Is there a way to test async code with miri without it giving and error at the 'wrong' place. Like saying ignore async unsoudness and just look at my code, if that makes sense.

**Reply by Darksonn** [reply]
Noalias is irrelevant when it comes to miri since its an LLVM specific thing. Generally, there's no way around this at the moment — miri is just bad at evaluating async code.

**Reply by soerenmeier** [reply]
Okay, hopefully that will change in the future. Thanks for the help.
